### ISSUE TITLE
On-line analysis 

### ISSUE THREAD

**Issue by yanivbi** [CyberReboot/NetworkML#75]
Hi,
Is is possible to use PoseidonML for on-line traffic analysis? 
i.e., instead of use pcap(s) as input, use tcpdump output as input to PoseidonML.

**Comment by gregs5** [comment]
I'm not sure I fully understand the question, but in an attempt to help answer: Poseidon takes raw traffic as input, and extracts a number of features from that traffic. It does this in 15min increments. (i.e. It looks at 15min worth of traffic per "chunk" of data) That 15min of traffic could be from 15min ago, or 2 weeks ago - Poseidon doesn't care, as from its perspective its just looking at raw data.  We usually use tcpdump to gather that raw traffic, but one could use other tools, for sure.

Is that what you are asking? Or are you asking if Poseidon can use the text format output of tcpdump? (If the latter, no - at this stage it requires the raw traffic - at least the headers, anyway)  Does this help?

**Comment by yanivbi** [comment]
Let me rephrase the question. 
I would like to "look" at the traffic from now and on, and not on previously collected traffic.
So, instead of using already collected pcap as an input, I would like to indicated the interface to which it should look into, for instance, by tcpdump.
Is it more clear that way? 
Thank you.

**Comment by yanivbi** [comment]
Hi,
I have seen https://github.com/CyberReboot/poseidon
but it's designed to support SDN-based networks, meaning that I need to integrate poseidon with a specific controller, right? 
In my case, I would like to run poseidon as standalone and let it look in the ingress traffic to a specific interface in a "streaming analytic" manner.

BTW, what is the right way to use Poseidon ML? 
train, and then eval or testing? what's the purpose of each? 

Thank you,

**Comment by cglewis** [comment]
Hi @yanivbi 

For the right way to use Poseidon ML regarding training/evaluating/testing I'd defer to @CStephenson970.

Yes, Poseidon is designed to support SDN networks.  Behind the scenes Poseidon uses Vent, which is how PoseidonML plugs in.

Vent provides an `ncontrol` plugin that can be used to start a tap on a live NIC which will write out PCAPs from a live network at intervals specified in the payload (http://vent.readthedocs.io/en/latest/core_and_plugins.html#network-tap).

So, I'd recommend trying out Vent (https://github.com/CyberReboot/vent) and adding PoseidonML as a plugin.  Also, the way that PoseidonML uses PCAPs it expects sessions to be cut up, so we also use this plugin to do that for us: https://github.com/CyberReboot/vent-plugins/tree/master/pcap_to_node_pcap.

**Comment by gregs5** [comment]
re: PoseidonML training, @CStephenson970 is the authority on that side, but I can share what *I* do: typically I'll get a collection of pcaps together that I know / have labelled - usually several days worth - and then I'll follow the training process from the README.  I'll hold a few of them back to test that my model is working once trained, but my rough process for the device/role identification is:

a) collect and label (pcaps)
b) train
c) test

This, btw, is also where the idea came for https://github.com/CyberReboot/pcapplot;  I had some not-ideal pcaps that were skewing my results, and I needed a better way of spotting the "off" ones.  It's a bit of trial and error, but as a non-ML guy I've found the process to be pretty smooth once you know the sequence / understand the scripts.  

And +1 on @cglewis : vent suggestion.  It would be good to know what your results/experiences are on that front, as I'm sure you're not the only person who doesn't want SDN as their starting point.  Hope this helps!

**Comment by cglewis** [comment]
regarding the tap on the live NIC - Vent will do that bit via the `ncontrol` plugin.  That tap will write out to PCAP files as it's capturing live, which `pcap_to_node_pcap` will then cut up, and then be picked up by PoseidonML.  So it's not processing the live NIC in realtime, but the ML models need the 15 minute chunk of traffic anyway, so it doesn't really matter that it's being written out to file first, in this case.

**Comment by gregs5** [comment]
Hey @yanivbi - what we've been doing re: pcaps is capturing off our internal networks, and then using that data to train. We've shared trained models externally (an earlier version is bundled w/ PoseidonML currently, for example), and we'll push a more updated one up in the coming weeks.  That said, we've been very interested in mixing our sanitized pcaps with pcaps from a different network, but it's hard to find good partners for this. Some background, if you're curious:

https://blog.cyberreboot.org/tcpdump-and-the-care-and-feeding-of-an-intelligent-sdn-eca6e7506342 

Would you be willing to share sanitized pcaps if we shared ours?

**Comment by gregs5** [comment]
Oh, and as for labelling pcaps, yes; for label_assignments.json.  At the risk of stating the obvious: when we are collecting off of a span or mirror port that doesn't have a single host on it, we need to "pull out" the traffic for an individual node.  For example, I might have 20 nodes represented in the initial pcap, and I'll want to pull out ones that I want to use for training - one at a time - and create individual pcaps for them.  So I'll have a Windows_Desktop one, or a Mac_DeveloperWorkstation, or a Sharepoint_Server, etc.  Once I have the pcaps of individual nodes all organized, that's when I modify the label assignments json, and start the ML training.

Does this help?

**Comment by jspeed-meyers** [comment]
This conversation has ended and there does not appear to be an issue.
